File: Data Structures & Algorithms/binary-tree-inorder-traversal/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        // same as pre-order traversal in here -> first go to left subtree , then Node , then go the right subtree
        List<Integer> ans = new ArrayList<>();
        // now call the hlper fucntion
        inOrder(root,ans);
        // after getting it
        return ans;
    }
    // helper function
    public void inOrder(TreeNode root , List<Integer> ans){
        //base important case of this recursive fucntion -> if the current node is null go back to its previous recursive called fucntion
        if(root == null){
            return;
        }
        // in inorder traversal -> first left subtree we go and print it
        inOrder(root.left,ans);
        // then thats Node value (ancestor)
        ans.add(root.val);
        // now go to its right side
        inOrder(root.right,ans);
    }
}

File: Data Structures & Algorithms/balanced-binary-tree/submission-0.cpp
#include <bits/stdc++.h>
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isBalanced(TreeNode* root) {
         // If our bottom-up helper function returns -1, the tree is unbalanced
        return dfsHeight(root) != -1;
    }

    int dfsHeight(TreeNode* root){

        // sicne its a recurive ficntion very imporat base case
        if(root == nullptr){
            return 0;
        }

        int leftHeight = dfsHeight(root->left);
        // after calculating the height of left side subtree if its equal to -1 return it
        if(leftHeight == -1){
            return -1; // Abort: Left side is already unbalanced
        }
        // now after left side -> now go towards right side
        int rightHeight = dfsHeight(root->right);
        if(rightHeight == -1){
            return -1;  // Abort: right side is already unbalanced
        }

        // 4. Evaluate current parent node using the heights passed up

        // after cacluting both : very imporant step formula -> if differnence between those 2 are > 1 then they r not balanced tree -> immediatl return -1
        if(abs(leftHeight - rightHeight) > 1){
            return -1;   // Abort: Current node breaks the balance condition
        }

        // now calculate its max height ->  // 5. Node-Counting Induction: Return actual height if perfectly balanced
        return max(leftHeight,rightHeight) + 1; // here we r following node based height counting
    }
};